src/Groups/GAPGroups.jl

```diff
@@ -0,0 +1,617 @@
+# further possible functions: similar, literal_pow, parent_type
+
+export order, perm, cperm, isfinite, gens, ngens, comm, comm!, inv!, rand_pseudo, one!, div_right,
+div_left, div_right!, div_left!, deg, mul, mul!, listperm, degree, elements, right_coset, coset_decomposition,
+right_cosets , right_transversal, conjugacy_class, conjugacy_classes, number_conjugacy_classes, isconjugate, conjugacy_classes_subgroups, conjugacy_classes_maximal_subgroups, normalizer, normaliser, core, pcore, fitting_subgroup, frattini_subgroup, radical_subgroup, socle, sylow_subgroup, hall_subgroup, sylow_system, complement_system, hall_system, relators, isperfect, issimple, isalmostsimple, ispgroup
+
+
+# TODO: as soon as GAP packages like `polycyclic` or `rcwa` are loaded,
+# the custom group types and isos they define should be added to the arrays
+# _gap_group_types resp. _iso_function
+
+
+function group_element(G::T, x::GapObj) where T <: Group
+  return GAPGroupElem{T}(G, x)
+end
+
+function elements(G::T) where T <: Group
+  els = GAP.gap_to_julia(GAP.Globals.Elements(G.X))
+  elems = Vector{elem_type(G)}(undef, length(els))
+  i = 1
+  for x in els
+    elems[i] = group_element(G, x)
+    i += 1
+  end
+  return elems
+end
+
+# to be fixed later
+Base.hash(x::GAPGroupElem) = 0
+
+function parent(x::GAPGroupElem)
+  return x.parent
+end
+
+function Base.isfinite(G::PermGroup)
+  return true
+end
+
+function Base.isfinite(G::PcGroup)
+  return true
+end
+
+function Base.isfinite(G::Group)
+  return GAP.Globals.IsFinite(G.X)
+end
+
+function degree(x::PermGroup)
+   return x.deg
+end
+
+function order(x::Group)
+   return GAP.gap_to_julia(GAP.Globals.Size(x.X))
+end
+
+function order(x::GAPGroupElem)
+   return GAP.gap_to_julia(GAP.Globals.Order(x.X))
+end
+
+function order(::Type{T}, x::Union{GAPGroupElem, Group}) where T<:Number
+   return T(order(x))
+end
+
+Base.exponent(x::Group) = GAP.Globals.Exponent(x.X)
+
+Base.length(x::Group) = order(x)
```

Comment:
Is this a good idea? The only justification I can think of is that perhaps it is required by Julia's `iterate`? @GDeFranceschi please check the Julia documentation for that. If it is needed for `iterate`, let's move this function next to the `iterate` code; else remove it.

Reply:
Yes iterators are assumed to implement `Base.length`, by default. It's possible to opt-out by implementing the `IteratorSize` trait, e.g. `Base.IteratorSize(::Type{<:Group}) = Base.SizeUnknown()`.

Reply:
Yeah, as I wrote below: for PermGroup and PcGroup, we know they are always finite, so this works. But for e.g. a non-trivial free group, that's infinite, so what should `length` return? Probably needs to throw an exception. So, in summary, here is what I suggest (and @GDeFranceschi should then implement it):

1. add `Base.IteratorSize(::Type{<:Group}) = Base.SizeUnknown()` so that by default, we assume groups can be infinite
2. add `Base.IteratorSize(::Type{PermGroup}) = HasLength()` etc. for types where we know they are finite (in theory we could also cover types that we know are *always* infinite, but I am not sure any exist in GAP: you almost always can represent the trivial group with any "type" of groups)
3. move this `length` method to sit next to the rest of the `iterator` interface
4. change it to cast its result to `Int`, so that for groups that are infinite, or just happen to contain more than 2^64 elements, we get a suitable `InexactError`
5. Add test cases covering all these, i.e., iterating over trivial groups of all "types"; iterating over some non-trivial but finite groups; and perhaps also one trying to iterate over an infinite one (for just a few steps, of course)

Reply:
I did all the points, but the behaviour of the iterator does not change. If I try a for cycle on an infinite group (e.g. a free_group), it starts running infinitely and I have to force Julia to quit.

Reply:
It does not work. If I type `[y for y in G]`, it returns the error message `HasLength not defined`.

Reply:
>  it returns the error message HasLength not defined.

I guess you wrote `HasLength()` without importing it nor qualifying it: if that's the case, write `Base.HasLength()`.

Reply:
yeah, well, if you do `[y for y in G]` for an infinite (or even merely large, say size 2^50) group, then I'd indeed expect it to run essentially infinitely long: if you really want to enter an infinite loop, the computer won't stop you. You might have your reasons after all (e.g. perhaps the iterator eventually raises an exception -- Julia can't know)